saml/src/main/java/com/linecorp/armeria/server/saml/SamlMetadataServiceFunction.java

```diff
@@ -0,0 +1,208 @@
+import com.linecorp.armeria.common.HttpData;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+/**
+ * A service which returns the SAML metadata in response to a user request.
+ */
+final class SamlMetadataServiceFunction implements SamlServiceFunction {
+    private static final Logger logger = LoggerFactory.getLogger(SamlMetadataServiceFunction.class);
+
+    @VisibleForTesting
+    static final MediaType CONTENT_TYPE_SAML_METADATA = MediaType.parse("application/samlmetadata+xml");
+
+    private static final HttpHeaders HTTP_HEADERS =
+            HttpHeaders.of(HttpStatus.OK)
+                       .contentType(CONTENT_TYPE_SAML_METADATA)
+                       .add(HttpHeaderNames.CONTENT_DISPOSITION,
+                            "attachment; filename=\"saml_metadata.xml\"")
+                       .asImmutable();
+
+    private final String entityId;
+    private final Credential signingCredential;
+    private final Credential encryptionCredential;
+    private final Map<String, SamlIdentityProviderConfig> idpConfigs;
+    private final Collection<SamlAssertionConsumerConfig> assertionConsumerConfigs;
+    private final Collection<SamlEndpoint> singleLogoutEndpoints;
+
+    private final ConcurrentMap<String, HttpData> metadataMap = new MapMaker().makeMap();
+
+    SamlMetadataServiceFunction(String entityId,
+                                Credential signingCredential,
+                                Credential encryptionCredential,
+                                Map<String, SamlIdentityProviderConfig> idpConfigs,
+                                Collection<SamlAssertionConsumerConfig> assertionConsumerConfigs,
+                                Collection<SamlEndpoint> singleLogoutEndpoints) {
+        this.entityId = entityId;
+        this.signingCredential = signingCredential;
+        this.encryptionCredential = encryptionCredential;
+        this.idpConfigs = idpConfigs;
+        this.assertionConsumerConfigs = assertionConsumerConfigs;
+        this.singleLogoutEndpoints = singleLogoutEndpoints;
+    }
+
+    @Override
+    public HttpResponse serve(ServiceRequestContext ctx, AggregatedHttpMessage msg,
+                              String defaultHostname, SamlPortConfig portConfig) {
+        final HttpData metadata = metadataMap.computeIfAbsent(defaultHostname, h -> {
+            try {
+                final Element element =
+                        SamlMessageUtil.marshalMessage(buildMetadataEntityDescriptorElement(h, portConfig));
+                final HttpData newMetadata = HttpData.ofUtf8(nodeToString(element));
+                logger.debug("SAML service provider metadata has been prepared for virtual host {}.", h);
```

Comment:
How about: `for: {}`, because VirtualHost.toString() starts with `VirtualHost(`

Reply:
Good point! Fixed.